tools/replay.c

```diff
@@ -0,0 +1,830 @@
+				continue;
+
+			if (gathers[i].id != relocs[k].gather_id)
+				continue;
+
+			rbo = lookup_bo(relocs[k].id, relocs[k].ctx_id);
+			assert(rbo != NULL);
+
+			pb->ptr = pb->bo->ptr + relocs[k].patch_offset;
+
+			ret = HOST1X_PUSHBUF_RELOCATE(pb, rbo->bo,
+						      relocs[k].offset, 0);
+			assert(ret == 0);
+
+			handled_relocs[k] = true;
+		}
+
+		pb->offset = gathers[i].offset;
+		pb->length = gathers[i].num_words;
+	}
+
+	for (i = 0; i < num_relocs; i++) {
+		char *fbsrt = "";
+
+		if (!rfb) {
+			rfb = lookup_framebuffer(relocs[i].id, relocs[i].ctx_id);
+			if (rfb)
+				fbsrt = "(fb)";
+		} else if (lookup_framebuffer(relocs[i].id, relocs[i].ctx_id)) {
+			fbsrt = "(fb)";
+		}
+
+		printf("        reloc %u bo_id: %u ctx_id: %u offset: %u %s\n",
+		       i, relocs[i].id, relocs[i].ctx_id, relocs[i].offset, fbsrt);
+		assert(relocs[i].ctx_id == relocs[0].ctx_id);
+		assert(handled_relocs[i]);
+	}
+
+	job->syncpt_incrs = num_syncpt_incrs;
+
+	ret = HOST1X_CLIENT_SUBMIT(client, job);
+	if (ret < 0)
+		abort();
+
+	host1x_job_free(job);
+
+	ret = HOST1X_CLIENT_FLUSH(client, &fence);
+	if (ret < 0)
+		abort();
+
+	ret = HOST1X_CLIENT_WAIT(client, fence, ~0u);
+	if (ret < 0)
+		abort();
+
+	if (rfb)
+		display_framebuffer(rfb);
+
+	return 0;
+}
+
+static void handle_single_step(void)
+{
+	static bool ss = false;
+
+	if (ss || grate_key_pressed2(NULL) == 32 /* Space */) {
+		fprintf(stderr, "Press Space to single step or Enter to continue\n");
+
+		while (1) {
+			uint8_t key = grate_key_pressed2(NULL);
+
+			if (key == 32 /* Space */) {
+				ss = true;
+				break;
+			}
+
+			if (key == 10 /* Enter */) {
+				ss = false;
+				break;
+			}
+		}
+	} else {
+		ss = false;
+	}
+}
+
+int main(int argc, char *argv[])
+{
+	struct record_act r;
+	unsigned int act_cnt = 0;
+	char *path = NULL;
+	int err = 1;
+	int ret;
+	int c;
+
+	host1x = host1x_open(true, -1);
+	if (!host1x) {
+		fprintf(stderr, "host1x_open() failed\n");
+		abort();
+	}
+
+	display = host1x_get_display(host1x);
+	if (display) {
+		ret = host1x_overlay_create(&overlay, display);
+		if (ret < 0) {
+			fprintf(stderr, "host1x_overlay_create() failed %d\n",
+				ret);
+			abort();
+		}
+	}
+
+	gr2d = host1x_get_gr2d(host1x);
+	if (!gr2d) {
+		fprintf(stderr, "host1x_get_gr2d() failed\n");
+		abort();
+	}
+
+	gr3d = host1x_get_gr3d(host1x);
+	if (!gr3d) {
+		fprintf(stderr, "host1x_get_gr3d() failed\n");
+		abort();
+	}
+
+	do {
+		struct option long_options[] =
+		{
+			{"recfile",	required_argument, NULL, 0},
+			{ /* Sentinel */ }
+		};
+		int option_index = 0;
+
+		c = getopt_long(argc, argv, "h", long_options, &option_index);
+
+		switch (c) {
+		case 0:
+			switch (option_index) {
+			case 0:
+				path = optarg;
+				break;
+
+			default:
+				return 0;
+			}
+
+		case -1:
+			break;
+
+		default:
+			fprintf(stderr, "Invalid arguments\n\n");
+			abort();
+		}
+	} while (c != -1);
+
+	if (!path) {
+		fprintf(stderr, "'--record path' is missing\n\n");
```

Comment:
I think this should be `--recfile` instead of `--record`, no?

Reply:
Yes!